src/ray/object_manager/test/pull_manager_test.cc

```diff
@@ -0,0 +1,158 @@
+
+#include "ray/object_manager/pull_manager.h"
+
+#include "gtest/gtest.h"
+#include "ray/common/test_util.h"
+
+namespace ray {
+
+class PullManagerTest : public ::testing::Test {
+ public:
+  PullManagerTest()
+      : self_node_id_(NodeID::FromRandom()),
+        object_is_local_(false),
+        num_send_pull_request_calls_(0),
+        rand_int_(0),
+        num_restore_spilled_object_calls_(0),
+        pull_manager_(self_node_id_,
+                      [this](const ObjectID &object_id) { return object_is_local_; },
+                      [this](const ObjectID &object_id, const NodeID &node_id) {
+                        num_send_pull_request_calls_++;
+                      },
+                      [this](int) { return rand_int_; },
+                      [this](const ObjectID &, const std::string &,
+                             std::function<void(const ray::Status &)>) {
+                        num_restore_spilled_object_calls_++;
+                      }) {}
+
+  NodeID self_node_id_;
+  bool object_is_local_;
+  int num_send_pull_request_calls_;
+  int rand_int_;
+  int num_restore_spilled_object_calls_;
+  PullManager pull_manager_;
+};
+
+TEST_F(PullManagerTest, TestStaleSubscription) {
+  ObjectID obj1 = ObjectID::FromRandom();
+  rpc::Address addr1;
+  ASSERT_EQ(pull_manager_.NumActiveRequests(), 0);
+  pull_manager_.Pull(obj1, addr1);
+  ASSERT_EQ(pull_manager_.NumActiveRequests(), 1);
+
+  std::unordered_set<NodeID> client_ids;
+  pull_manager_.OnLocationChange(obj1, client_ids, "");
+
+  // There are no client ids to pull from.
+  ASSERT_EQ(num_send_pull_request_calls_, 0);
+  ASSERT_EQ(num_restore_spilled_object_calls_, 0);
+
+  client_ids.insert(NodeID::FromRandom());
+  pull_manager_.OnLocationChange(ObjectID::FromRandom(), client_ids, "");
+
+  // Now we're getting a notification about an object we aren't trying to pull.
+  ASSERT_EQ(num_send_pull_request_calls_, 0);
+  ASSERT_EQ(num_restore_spilled_object_calls_, 0);
+
+
+  pull_manager_.CancelPull(obj1);
+  ASSERT_EQ(pull_manager_.NumActiveRequests(), 0);
+}
+
+TEST_F(PullManagerTest, TestRestoreSpilledObject) {
+  ObjectID obj1 = ObjectID::FromRandom();
+  rpc::Address addr1;
+  ASSERT_EQ(pull_manager_.NumActiveRequests(), 0);
+  pull_manager_.Pull(obj1, addr1);
+  ASSERT_EQ(pull_manager_.NumActiveRequests(), 1);
+
+  std::unordered_set<NodeID> client_ids;
+  pull_manager_.OnLocationChange(obj1, client_ids, "remote_url/foo/bar");
+
+  // client_ids is empty here, so there's nowhere to pull from.
+  ASSERT_EQ(num_send_pull_request_calls_, 0);
+  ASSERT_EQ(num_restore_spilled_object_calls_, 1);
+
+  client_ids.insert(NodeID::FromRandom());
+  pull_manager_.OnLocationChange(obj1, client_ids, "remote_url/foo/bar");
+
+  // The behavior is supposed to be to always restore the spilled object if possible (even if it exists elsewhere in the cluster).
+  ASSERT_EQ(num_send_pull_request_calls_, 0);
+  ASSERT_EQ(num_restore_spilled_object_calls_, 2);
```

Comment:
@rkooo567 this is a little weird right? Once the object is in the cluster again, I would've thought it would make more sense to pull it from plasma than an external store.

Reply:
Currently, we always pull the object from the ext storage once it is spilled (for the simple design for MVP). I agree with you it is weird, and this make fs ext storage not work in multi-node setup (https://github.com/ray-project/ray/issues/12044), so we will likely to fix in the future. For now, you can probably add TODO there (maybe TODO(sang)?)